### DISCUSSION TITLE
Unsure how to generate all satisfying models

### DISCUSSION THREAD

**Discussion by fedelebron** [bitwuzla/bitwuzla#14] [General] [answered]
Hi!

I have a `.btor` with the following contents:

```
1 var 64
2 constd 6 18
3 sll 64 1 2
4 add 64 -1 3
5 constd 6 31
6 srl 64 4 5
7 xor 64 4 6
8 constd 64 21
9 mul 64 7 8
10 constd 6 11
11 srl 64 9 10
12 xor 64 9 11
13 constd 6 6
14 sll 64 12 13
15 add 64 12 14
16 constd 6 22
17 srl 64 15 16
18 xor 64 15 17
19 consth 64 ffffffff
20 and 64 19 18
21 consth 64 deadbeef
22 eq 1 20 21
23 root 1 22
```

I'd like to generate all possible satisfying assignments of the bitvector 1. If I call bitwuzla, I get:

```
$ bin/bitwuzla ~/chacha20/mowhash.btor -x --btor -m
sat
1 46d2694a4431f014
```

I tried using `-i` for incremental solving (which should be somehow related to finding all solutions, not just one), but it's not clear to me how this incremental mode works, I was unable to find documentation about it.

How, if at all possible, can I ask bitwuzla to generate all possible assignments of the bitvector 1?

**Comment by recoules** [comment, marked as answer]
Hi @fedelebron,

I am not a `bitwuzla` developper, so it may exist a more efficient way to do it, but here is a way to do *this*.  
Solve the formula the first time, get the satisfying assignment of the bitvector and add a new constraint to the formula such that the bitvector is distinct of this value. Solve the formula again and repeat the process until the formula become unsatisfiable.  
This is generally done programmatically (`C`, `python` or `OCaml` API).

Still, my though is that SMT solvers are not really designed for exhaustive solution enumeration. 

For example, with the `smt2` syntax (I am not used to `btor` syntax), it could be:
```smt2
(declare-const x (_ BitVec 8))
(assert (bvult x #x03))
(check-sat)
```
```console
sat
```
```smt2
(get-model)
```
```console
(
  (define-fun x () (_ BitVec 8) #x00)
)
```
```smt2
(assert (not (= x #x00)))
(check-sat)
```
```console
sat
```
```smt2
(get-model)
```
```console
(
  (define-fun x () (_ BitVec 8) #x02)
)
```
```smt2
(assert (not (= x #x02)))
(check-sat)
```
```console
sat
```
```smt2
(get-model)
```
```console
(
  (define-fun x () (_ BitVec 8) #x01)
)
```
```smt2
(assert (not (= x #x01)))
(check-sat)
```
```console
unsat
```

**Reply by fedelebron** [reply]
Thank you! I'll give that a try :)
